packages/core/src/new-api/internal/execution/chain-dispatcher.ts

```diff
@@ -0,0 +1,160 @@
+
+import { Adapters } from "../../types/adapters";
+import { ArgumentType } from "../../types/module";
+import { ChainDispatcher } from "../types/chain-dispatcher";
+import { assertIgnitionInvariant } from "../utils/assertions";
+
+import { AccountsState } from "./execution-engine";
+
+export class ChainDispatcherImpl implements ChainDispatcher {
+  private _accountsState: AccountsState;
+
+  constructor(private _adapters: Adapters) {
+    this._accountsState = {};
+  }
+
+  // TODO: should nonce management be separated out into its own
+  // class?
+  public async allocateNextNonceForAccount(address: string): Promise<number> {
+    if (address in this._accountsState) {
+      const nextNonce = this._accountsState[address] + 1;
+      this._accountsState[address] = nextNonce;
+      return nextNonce;
+    }
+
+    const onchainNonce = await this._getTransactionCount(address);
+    this._accountsState[address] = onchainNonce;
+
+    return onchainNonce;
+  }
+
+  public async constructDeployTransaction(
+    byteCode: string,
+    abi: any[],
+    args: ArgumentType[],
+    value: bigint,
+    from: string
+  ): Promise<ethers.providers.TransactionRequest> {
+    const signer: ethers.Signer = await this._adapters.signer.getSigner(from);
+
+    const Factory = new ContractFactory(abi, byteCode, signer);
+
+    const tx = Factory.getDeployTransaction(...args, {
+      value,
+    });
+
+    return tx;
+  }
+
+  public async constructCallTransaction(
+    contractAddress: string,
+    abi: any[],
+    functionName: string,
+    args: ArgumentType[],
+    value: bigint,
+    from: string
+  ): Promise<ethers.providers.TransactionRequest> {
+    const signer: ethers.Signer = await this._adapters.signer.getSigner(from);
+
+    const contractInstance = new Contract(contractAddress, abi, signer);
+
+    const unsignedTx: ethers.providers.TransactionRequest =
+      await contractInstance.populateTransaction[functionName](...args, {
+        value: BigInt(value),
+        from,
+      });
+
+    return unsignedTx;
+  }
+
+  public async sendTx(
+    tx: ethers.providers.TransactionRequest,
+    from: string
+  ): Promise<string> {
+    const signer = await this._adapters.signer.getSigner(from);
+    const response = await signer.sendTransaction(tx);
+    const txHash = response.hash;
+
+    return txHash;
+  }
+
+  public async getEventArgument(
+    eventName: string,
+    argumentName: string,
+    txToReadFrom: string,
+    eventIndex: number,
+    emitterAddress: string,
+    abi: any[]
+  ): Promise<any> {
+    const contract = new Contract(emitterAddress, abi);
+    const filter = contract.filters[eventName]();
+    const eventNameTopic = filter.topics?.[0];
+
+    assertIgnitionInvariant(eventNameTopic !== undefined, "Unknown event name");
+
+    const receipt = await this._adapters.transactions.getTransactionReceipt(
+      txToReadFrom
+    );
+
+    // TODO: should this really return an error result
+    assertIgnitionInvariant(
```

Comment:
Is this one instance of the transaction dropped case?

Reply:
We are happy with assertion here, this should never be called on an unconfirmed transaction.